File: svm/cpp/svm.cpp
#include <iostream>
#include <vector>

using namespace std;

class SVM {
 public:
  SVM(int C, char *kernel) {
    this->C = C;
    this->kernel = kernel;
    this->w = new vector<double>(2);
    this->b = 0.0;
  }

  void fit(vector<double> *X, vector<int> *y, int n) {
    for (int i = 0; i < n; i++) {
      double score = 0.0;
      for (int j = 0; j < 2; j++) {
        score += this->w[j] * X[i][j];
      }
      if (y[i] * score + this->b <= 1.0) {
        for (int j = 0; j < 2; j++) {
          this->w[j] += this->C * y[i] * X[i][j];
        }
        this->b += this->C * y[i];
      }
    }
  }

  int predict(vector<double> *X) {
    double score = 0.0;
    for (int j = 0; j < 2; j++) {
      score += this->w[j] * X[j];
    }
    return score >= 0.0 ? 1 : -1;
  }

  ~SVM() {
    delete this->w;
  }

 private:
  int C;
  char *kernel;
  vector<double> *w;
  double b;
};

File: svm/rust/svm.rs
use std::f32;

struct SVM {
    weight: Vec<f32>,
    bias: f32,
}

impl SVM {
    fn new(c: f32, kernel: &str) -> SVM {
        SVM {
            weight: vec![0.0; 2],
            bias: 0.0,
        }
    }

    fn fit(&mut self, x: &[Vec<f32>], y: &[i32], n: usize) {
        for _ in 0..n {
            let score = self.weight.dot(&x[0]) + self.bias;
            if y[0] * score <= 1.0 {
                for i in 0..2 {
                    self.weight[i] += c * y[0] * x[0][i];
                }
                self.bias += c * y[0];
            }
        }
    }

    fn predict(&self, x: Vec<f32>) -> i32 {
        let score = self.weight.dot(&x) + self.bias;
        if score >= 0.0 { 1 } else { -1 }
    }
}

fn main() {
    let svm = SVM::new(1.0, "linear");
    svm.fit(&[[1.0, 2.0], [3.0, 4.0]], &[1, -1], 2);
    println!("{}", svm.predict([5.0, 6.0])); // prints 1
}
